source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -259,6 +259,57 @@ Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
   return repeatReadAndUnprotect(buffer, result);
 }
 
+Network::IoResult TsiSocket::repeatProtectAndWrite(Buffer::Instance& buffer, bool end_stream) {
+  uint64_t total_bytes_written = 0;
+  Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
+
+  while (true) {
+    uint64_t bytes_to_drain_this_iteration =
```

Comment:
There's an odd case that merits special consideration:

`prev_bytes_to_drain_ == 0 && raw_write_buffer_.length() > 0`

I think this can happen in the case where doHandshake adds bytes to raw_write_buffer_ but also completes the handshake.  When this happens, the protect call below is skipped, but bytes_to_drain_this_iteration ends up being > 0, which could result in bytes in the input buffer being discarded without being sent.

Ways to detect:
`ASSERT((prev_bytes_to_drain_ == 0) == (raw_write_buffer_.length() == 0);`
`ASSERT(prev_bytes_to_drain_ >= buffer.length());`
`ASSERT(buffer.length() >= bytes_to_drain_this_iteration)` before the call to `buffer.drain()` further down
Possibly adding an `ASSERT` to `OwnedImpl::drainImpl` to detect attempts to drain more bytes than are in the buffer.

Reply:
Per [this code](https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/alts/tsi_socket.cc#L168-L170), I think we have a guarantee that raw_write_buffer_.length() is 0 when entering doWrite() for the first time after handshake completes. Also, it does not make sense that a peer wants to send non-handshake data without first confirming if the handshake completes successfully. In other words, during handshake, peer A will send whatever it receives from peer B to the handshake service in order to get the bytes to send to peer B. Here, peer A will not concatenate any non-handshake data to the data received from peer B, and send them to the handshake service because peer A has not received any confirmation from the handshake service that handshake completes successfully.

Reply:
raw_write_buffer_ will be non-empty after the doWrite just after handshake completes if that doWrite results in a partial write.  I know this may be really unlikely but it is possible for raw_write_buffer_ to be non-empty after handshake completes.  I think it also true that the peer that completes the handshake first will need to do a write after handshake completes locally to provide the remote peer the information it needs to complete its own handshake.

Reply:
Possible solution: branch on `raw_write_buffer_.length() > 0` instead of `prev_bytes_to_drain_`.

When `raw_write_buffer_.length() > 0` then `bytes_to_drain_this_iteration = prev_bytes_to_drain_` and we should attempt a write even if `bytes_to_drain_this_iteration` is 0 which would happen if the bytes in the buffer are handshake bytes.

When `raw_write_buffer_.length() > 0` then `bytes_to_drain_this_iteration = std::min(buffer.length(), max_unprotected_frame_size_)`.  If >0, attempt to protect and write those bytes, else break.

Reply:
I took a slightly different approach by introducing a new field - `prev_handshake_bytes_to_drain_` that indicates if we need to drain handshake data before doing regular protect+write operations. PTAL.